File: frontend/src/app/models/options-chain.ts
//
// Date: 4/4/2018
// Author(s): Spicer Matthews (spicer@options.cafe)
// Copyright: 2018 Cloudmanic Labs, LLC. All rights reserved.
//

import * as moment from 'moment';
import { Symbol } from './symbol';

//
// OptionsChain
//
export class OptionsChain 
{
  Underlying: string;
  ExpirationDate: Date;
  Calls: OptionsChainItem[];
  Puts: OptionsChainItem[];


  //
  // Json to Object.
  //
  fromJson(json: Object): OptionsChain 
  {
    let obj = new OptionsChain();

    obj.Underlying = json["underlying"];
    obj.ExpirationDate = moment(json["expiration_date"]).toDate();
    obj.Calls = [];
    obj.Puts = [];

    // Add in the calls.
    for (let i = 0; i < json["calls"].length; i++)
    {
      obj.Calls.push(new OptionsChainItem().fromJson(json["calls"][i]));
    }

    // Add in the puts.
    for (let i = 0; i < json["puts"].length; i++) {
      obj.Puts.push(new OptionsChainItem().fromJson(json["puts"][i]));
    }   

    return obj;
  }  
}

//
// OptionsChainItem
//
export class OptionsChainItem
{
  Underlying: string;
  Symbol: string;
  OptionType: string;
  Description: number;
  Strike: number;
  ExpirationDate: Date;
  Last: number;
  Change: number;
  ChangePercentage: number;
  Volume: number;
  AverageVolume: number;
  LastVolume: number;
  Open: number;
  High: number;
  Low: number;
  Close: number;
  Bid: number;
  BidSize: number;
  Ask: number;
  AskSize: number;
  OpenInterest: number;

  //
  // Json to Object.
  //
  fromJson(json: Object): OptionsChainItem 
  {
    let obj = new OptionsChainItem();
  
    obj.Underlying = json["underlying"];
    obj.Symbol = json["symbol"];
    obj.OptionType = json["option_type"];
    obj.Description = json["description"];
    obj.Strike = json["strike"];
    obj.ExpirationDate = moment(json["expiration_date"]).toDate();
    obj.Last = json["last"];
    obj.Change = json["change"];
    obj.ChangePercentage = json["change_percentage"];
    obj.Volume = json["volume"];
    obj.AverageVolume = json["average_volume"];
    obj.LastVolume = json["last_volume"];
    obj.Open = json["open"];
    obj.High = json["high"];
    obj.Low = json["low"];
    obj.Close = json["close"];
    obj.Bid = json["bid"];
    obj.BidSize = json["bid_size"];
    obj.Ask = json["ask"];
    obj.AskSize = json["ask_size"];
    obj.OpenInterest = json["open_interest"];

    return obj;
  }
}

/* End File */

File: backend/brokers/types/options_chain.go
//
// Date: 2018-04-03
// Author: Spicer Matthews (spicer@cloudmanic.com)
// Last Modified by: Spicer Matthews
// Last Modified: 2018-10-30
// Copyright: 2017 Cloudmanic Labs, LLC. All rights reserved.
//

package types

type OptionsChain struct {
	Id             uint               `gorm:"primary_key" json:"_"`
	Underlying     string             `sql:"not null" json:"underlying"`
	UnderlyingLast float64            `sql:"not null" json:"lunderlying_last"`
	ExpirationDate Date               `gorm:"type:date" sql:"not null" json:"expiration_date"`
	Puts           []OptionsChainItem `json:"puts"`
	Calls          []OptionsChainItem `json:"calls"`
}

type OptionsChainItem struct {
	OptionsChainId   uint    `sql:"not null;index:OptionsChainId" json:"_"`
	Underlying       string  `sql:"not null" json:"underlying"`
	Symbol           string  `sql:"not null" json:"symbol"`
	OptionType       string  `sql:"not null" json:"option_type"`
	Description      string  `sql:"not null" json:"description"`
	Strike           float64 `sql:"not null" json:"strike"`
	ExpirationDate   Date    `sql:"not null" json:"expiration_date"`
	Last             float64 `sql:"not null" json:"last"`
	Change           float64 `sql:"not null" json:"change"`
	ChangePercentage float64 `sql:"not null" json:"change_percentage"`
	Volume           int     `sql:"not null" json:"volume"`
	AverageVolume    int     `sql:"not null" json:"average_volume"`
	LastVolume       int     `sql:"not null" json:"last_volume"`
	Open             float64 `sql:"not null" json:"open"`
	High             float64 `sql:"not null" json:"high"`
	Low              float64 `sql:"not null" json:"low"`
	Close            float64 `sql:"not null" json:"close"`
	Bid              float64 `sql:"not null" json:"bid"`
	BidSize          int     `sql:"not null" json:"bid_size"`
	Ask              float64 `sql:"not null" json:"ask"`
	AskSize          int     `sql:"not null" json:"ask_size""`
	OpenInterest     int     `sql:"not null" json:"open_interest"`
	ImpliedVol       float64 `sql:"not null" json:"implied_vol""`
	Delta            float64 `sql:"not null" json:"delta""`
	Gamma            float64 `sql:"not null" json:"gamma""`
	Theta            float64 `sql:"not null" json:"theta""`
	Vega             float64 `sql:"not null" json:"vega""`
}

/* End File */
